Watch RBAC resources to trigger reconcile

kajinamit opened the pull request:
This ensures the controller watches service account, role, and role bindings to reconcile these resources in case any change is made.

stuggi commented on controllers/placementapi_controller_test.go:
```diff
@@ -27,143 +28,283 @@ import (
+// but does not create it yet. So the client can finetuned the Template data
+// before calling Create()
+func NewTestKeystoneAPI(namespace string) *TestKeystoneAPI {
+	name := fmt.Sprintf("keystone-%s", uuid.New().String())
+	template := &keystonev1.KeystoneAPI{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "keystone.openstack.org/v1beta1",
+			Kind:       "KeystoneAPI",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: TestNamespace,
+		},
+		Spec: keystonev1.KeystoneAPISpec{
+			DatabaseUser: "foo-bar-baz",
+		},
+		Status: keystonev1.KeystoneAPIStatus{
+			APIEndpoints:     map[string]string{"public": "fake-keystone-public-endpoint"},
+			DatabaseHostname: "fake-database-hostname",
+		},
+	}
+	return &TestKeystoneAPI{
+		LookupKey: types.NamespacedName{Name: name, Namespace: namespace},
+		Template:  template,
+		Instance:  &keystonev1.KeystoneAPI{},
+	}
+}
+
+// Creates the KeystoneAPI resource in k8s based on the Template. The Template
+// is not updated with the result of the create.
+func (t TestKeystoneAPI) Create() {
+	t.Instance = t.Template.DeepCopy()
+	Expect(k8sClient.Create(ctx, t.Instance)).Should(Succeed())
+
+	// the Status field needs to be written via a separate client
+	t.Instance.Status = t.Template.Status
+	Expect(k8sClient.Status().Update(ctx, t.Instance)).Should(Succeed())
+}
+
+// Deletes the KeystoneAPI resource from k8s and waits until it is deleted.
```
it might be useful to have this in a shared test module in lib-common. all services likely need an instance of keystone to be tested properly

gibizer replied:
That is a good idea.

stuggi replied:
I am working on finish the restructuring of lib-common. maybe we wait until that is landed. should be today or tomorrow I think

gibizer replied:
sure. I anyhow need to focus on OpenStack upstream for a while as feature freeze for Zed is 1st of Sept

stuggi reviewed (commented):
in general looks good to me, we could benefit from moving the keystone related helpers into a lib-common test module that it can be consumed buy other operators. probably all services required a keystone instance to be tested.

gibizer commented:
just rebased to the latest placement operator version

gibizer commented:
> in general looks good to me, we could benefit from moving the keystone related helpers into a lib-common test module that it can be consumed buy other operators. probably all services required a keystone instance to be tested.

This make sense. But first I would like to settle the move of the openshift CRD to lib-common( https://github.com/openstack-k8s-operators/lib-common/pull/26 ) as that also adds a test module there. Then I can start moving the keystone helpers to that test module

stuggi commented on controllers/placementapi_controller_test.go:
```diff
@@ -27,144 +28,284 @@ import (
+		client.InNamespace(t.LookupKey.Namespace),
+	}
+
+	err := k8sClient.List(ctx, configList, listOpts...)
```
not really an issue, just want to understand why we use here a CM list and loop over them to get the CM with the name name, instead of using Get() and query the specific CM? similar to https://github.com/openstack-k8s-operators/lib-common/blob/master/modules/common/configmap/configmap.go#L251

gibizer replied:
This show my inexperience with the framework from summer. When I did similar thing for nova recently[1] I did use Get() as it should be. 

All the ConfigMap, Job, Deployment and probably MariaDB and KeystoneEndpoint test helpers should move to lib-common/modules/test at some point. So let's keep this as is until it is replaced with better and common helper.

[1] https://github.com/openstack-k8s-operators/nova-operator/blob/5827164918d309565fb7290bb222092c6241a55b/test/functional/base_test.go#L181

stuggi reviewed (approved):
/lgtm

kajinamit committed:
Watch RBAC resources to trigger reconcile

This ensures the controller watches service account, role, and role
bindings to reconcile these resources in case any change is made.
